### ISSUE TITLE
MeeseeksDev extension to run benchmarks

### ISSUE THREAD

**Issue by jni** [scikit-image/scikit-image#3342]
@Carreau While I have your attention, I want to make a Meeseeks extension, `@meeseeksdev run benchmarks`, that will:
- check out the tip of a PR and run the asv benchmarks,
- check out the root of a PR and run the asv benchmarks,
- comment on the issue when said benchmarks are done with a summary of the benchmarks comparison.

I'll be at the limit of my comfort zone along many axes doing this so it would be good if we could coordinate a time with @TomAugspurger where both of you could be available on chat while I crank this out. Please let me know if this sounds worthwhile and I'll suggest a few times. Best days for me, always at 4pm Pacific:

- Wed 15
- Fri 17
- Mon 20
- Wed 22
- Fri 24

Or if you want to just do it yourself that also works. =P =P =P

**Comment by Carreau** [comment]
> Wed 15
> Fri 17

Those 2 dates are fine. The others I'm at JupyterCon. 


The one concern I have is that MrMeeseeks is on a heroku small instance. It likely does not have enough processing power to run ASV on scikit image. Though we can easily use that as a proxy that forward a command to another machine to run ASV and report back.

**Comment by TomAugspurger** [comment]
Wed 15th is better for me.

> Though we can easily use that as a proxy that forward a command to
another machine to run ASV and report back.

I think that's the plan. We're running the scikit-image on a bare-metal
pandas benchmark machine.

**Comment by Carreau** [comment]
> Wed 15th is better for me.

Good, 4pm pacific reserved on my calendar.

**Comment by jni** [comment]
Fantastic! Thank you both!

@Carreau, to confirm what @TomAugspurger said, the idea here is for Meeseeks to trigger these actions on a dedicated machine, not on its own. (Which in our case has already been set up, thanks Tom!)

Some ahead-of-time questions:

- @Carreau, when I write an extension, should it live on the Meeseeks repo, our repo, its own repo...?
- @TomAugspurger any docs I should read regarding triggering builds on airflow? Can I specify specific commits? What about merges of two commits?
- What about getting notified for the completed builds? Does ASV have nice "compare two commits" views rather than a history view?

Let's coordinate on gitter.im/scikit-image/scikit-image when the time comes.

Thanks again!

**Comment by jni** [comment]
For reference: https://github.com/scikit-image/scikit-image/pull/3137#issuecomment-396557700

**Comment by Carreau** [comment]
> when I write an extension, should it live on the Meeseeks repo, our repo, its own repo...?

Can be on the meeseeks repo, it's the easiest. It should be pretty simple. I can give you commit access to the repo, and to heroku to see the production logs. The hard part will be 1) authentication with the ASV server, and 2) having a hook/notification to trigger posting the response when the results are ready.

For 1) I think we can have a shared secret, and sign the requests... but we can discuss that Wednesday.

**Comment by Carreau** [comment]
> For 1) I think we can have a shared secret, and sign the requests... but we can discuss that Wednesday.

Maybe https://pypi.org/project/requests-http-signature/

**Comment by TomAugspurger** [comment]
https://github.com/teamclairvoyant/airflow-rest-api-plugin may be worth looking at.

Then the heroku server just has to worry about formatting an HTTP request, and all the logic will be on the ASV machine. I guess the heroku server will have to be listening for responses...

The tricky part will be defining an airflow task that responds to input. That should be doable, but will stretch my airflow knowledge. I'll look around.

**Comment by Carreau** [comment]
I may be able to poll, but one of the design of the heroku bit was to be state-less (or at least have no state it can't easily reconstruct).

It seem that heroku now have the equivalent of [free cron job](https://devcenter.heroku.com/articles/scheduler), so I might be able to squeeze that in my free tier, so we can still rely on polling.

**Comment by TomAugspurger** [comment]
Hopefully polling won't be necessary. Presumably the ASV server could make an HTTP request to the heroku bot with the right key & a payload containing the results?

**Comment by Carreau** [comment]
Yes it could, either using a personal access token, or Mr meeseeks can
provide a short lived token (up to 5min IIRC) that give limited access to
GitHub API, one specific repo, either when triggering asv (but I guess
token will expire before end of benchmark), or a simple API to fetch one.

My original plan with Mr meeseeks was to actually write a GitHub api proxy
that would hand of longer-lived token, with  more restrictions and do a
live translation. For example you hit https://meeseeks.com/ghprox/<GitHub
API path> but with a token that only allow comments on a specific PR for
example.

We can discuss this a this Afternoon meeting.

**Comment by TomAugspurger** [comment]
Do we have a link for the call?

**Comment by Carreau** [comment]
Same question https://appear.in/skimage ?

**Comment by jni** [comment]
So, you were both throwing around authentication schemes that I didn't understand in our chat, though I get most of it. Let me try to summarise the initial architecture:

- the bot receives a command from the PR (btw @Carreau how come mentions don't properly link to the bot? Put it another way: why not create the MeeseeksDev user to make it easier to tag?)
- the bot pings the ASV machine (URL specified by a config file on the skimage repo). Here, how does the ASV machine know it's a request from a valid source?
- the machine runs the benchmarks
- when the benchmarks are finished, the machine pings the bot again with specific info:
  * the PR number (which Meeseeks has forgotten)
  * some summary of the benchmarks (any benchmarks that have changed by > some threshold)
  * optionally, a link to the ASV html page. @TomAugspurger Is there a way to highlight the right commits in this view somehow?
  * Again, in this case how do we authenticate from Meeseeks that the ASV machine is a valid source of this request?
- Meeseeks comments on the PR with these results

**Comment by Carreau** [comment]
> (btw @Carreau how come mentions don't properly link to the bot? Put it another way: why not create the MeeseeksDev user to make it easier to tag?)

Because it's a "GitHub Application", that is to say a machine account. That's a github limitation. 
There are pro and cons for both, and the way the github UI react have change through time. You use to be able to mention bot with autocompletion. 
- Bot:
   - You can activate/deactivate a bot on a per-repository basis, no "accept" from the bot needed. - - - User:
   - You invite  and the user need to accept. 
- Bot:
   - You can request per-repo credential/token which is short lived (5 minutes)
- User: 
   - You use a personal access token that have access to everything....
- Bot:
   - can received webhook from GitHub
- User
   - Need to poll

Anyway, I know it's annoying but that's the way it is for now. There is [also a machine account](https://github.com/meeseeksmachine) That I use to submit PRs from Fork, which will allow me to not require Push permission to repo... Long story. 

The bot detect mention via regular expression, so I can also have other regex if you like. 

> the bot pings the ASV machine (URL specified by a config file on the skimage repo). Here, how does the ASV machine know it's a request from a valid source?

Or for now a config ENV Var on the bot. First easy step. Tom or you can set it via: 
$ heroku config:set SECRET=VALUE

If we want to do more complex, we can encrypt the secret on this repository with a pub key with the bot having the private one to decrypt. 

The bot ping ASV with the secret, and a payload to carry over, like  issue/pr number. 

> Again, in this case how do we authenticate from Meeseeks that the ASV machine is a valid source of this request?

Shared secret to sign request , all the communication are TLS so we are safe, but even if request leaks the secret is safe. 

I may suggest doing exactly what GH does, the code in Meeseks is [there](https://github.com/MeeseeksBox/MeeseeksDev/blob/master/meeseeksdev/meeseeksbox/core.py#L57-L62) and [there](https://github.com/MeeseeksBox/MeeseeksDev/blob/master/meeseeksdev/meeseeksbox/core.py#L57-L62)

There is a question of replay attack,  which we can mitigate by embedding  date/time/validity duration in the request. 

The rest seem fine to me.

**Comment by grlee77** [comment]
### comment moved from deleted Discussion thread

I think this was resolved instead by the GitHubActions workflow introduced by @jaimergp in #5424. We can now run benchmarks by adding the `run-benchmarks` tag to a PR.
